File: main/hw1-heap-sort/hw1-heap-sort.java
import java.util.Scanner;

final class Solution {
    public static void main(final String[] args) {
        final int[] a = readRecord();
        Algo.heapsort(a);
        print(a);
    }

    private static int[] readRecord() {
        try (final Scanner sc = new Scanner(System.in)) {
            final int[] a = new int[sc.nextInt()];
            for (int i = 0; i != a.length; ++i) a[i] = sc.nextInt();
            return a;
        }
    }

    private static void print(final int[] a) {
        System.out.print("[");

        String sep = "";
        for (final int x : a) {
            System.out.format("%s%d", sep, x);
            sep = ",";
        }

        System.out.println("]");
    }

    private Solution() { throw new AssertionError(); }
}

final class Algo {
    static void heapsort(final int[] a) {
        int stop = a.length;
        if (stop < 2) return;

        // Convert the array into a binary maxheap.
        for (int parent = (stop - 1) / 2; parent != -1; --parent)
            sift_down(a, a[parent], parent, stop);

        // Sort the array by successively popping elements from the maxheap.
        while (--stop != 0) {
            final int parent_elem = a[stop];
            a[stop] = a[0];
            sift_down(a, parent_elem, 0, stop);
        }
    }

    private static final int NO_CHILD = -1;

    private static void sift_down(final int[] a, final int parent_elem,
                                  int parent, final int stop) {
        for (; ; ) {
            final int child = pick_child(a, parent, stop);
            if (child == NO_CHILD || parent_elem >= a[child]) break;

            a[parent] = a[child];
            parent = child;
        }

        a[parent] = parent_elem;
    }

    private static int pick_child(final int[] a,
                                  final int parent, final int stop) {
        final int left = parent * 2 + 1;
        if (left >= stop) return NO_CHILD;

        final int right = left + 1;
        return right == stop || a[left] >= a[right] ? left : right;
    }

    private Algo() { throw new AssertionError(); }
}

File: main/hw1-heap-sort/hw1-heap-sort.cpp
#include <algorithm>
#include <iostream>
#include <iterator>
#include <utility>
#include <vector>

namespace {
    template<typename It>
    using Delta = typename std::iterator_traits<It>::difference_type;

    template<typename It>
    constexpr void heapsort(const It first, const It last) noexcept
    {
        constexpr Delta<It> no_child {-1};
        auto len = last - first;
        if (len == 0) return;

        const auto pick_child = [first, &len](const Delta<It> parent) noexcept {
            const auto left = parent * 2 + 1;
            if (left >= len) return no_child;

            const auto right = left + 1;
            return right == len || first[left] >= first[right] ? left : right;
        };

        const auto sift_down = [first, pick_child](Delta<It> parent) noexcept {
            auto elem = std::move(first[parent]);

            for (; ; ) {
                const auto child = pick_child(parent);
                if (child == no_child || elem >= first[child]) break;

                first[parent] = std::move(first[child]);
                parent = child;
            }

            first[parent] = std::move(elem);
        };

        // Convert the range into a binary maxheap.
        for (auto parent = len / 2 - 1; parent != -1; --parent)
            sift_down(parent);

        // Build a sorted range right-to-left by popping values from the heap.
        while (--len != 0) {
            std::iter_swap(first, first + len);
            sift_down(0);
        }
    }

    std::vector<int> read_record()
    {
        std::vector<int>::size_type n {};
        std::cin >> n;

        std::vector<int> a (n);
        for (auto& x : a) std::cin >> x;
        return a;
    }

    void print(const std::vector<int>& a)
    {
        std::cout << '[';

        auto sep = "";
        for (const auto x : a) {
            std::cout << sep << x;
            sep = ",";
        }

        std::cout << "]\n";
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    auto a = read_record();
    heapsort(begin(a), end(a));
    print(a);
}
